### ISSUE TITLE
Softmax for multiple dimensions

### ISSUE THREAD

**Issue by RobRomijnders** [tensorflow/tensorflow#3101]
Attention models get more popular. In many of my experiments with attention, I end up using Softmax over multiple dimensions. 

tf.nn.softmax() supports only one dimension. It would be nice to expand the API with another argument, softmax_dim, for softmax over multiple dimension, say over 2 (images) or 3 (CT-scans, videos)

I;d like to here what you think? DO more people end up writing softmax code themselves?

**Comment by andydavis1** [comment]
I'm not clear if this is a feature request, or question. Perhaps you could start by asking this question on StackOverflow first, and tag it with the `tensorflow` tag.

**Comment by vrv** [comment]
I think it's a feature request

**Comment by RobRomijnders** [comment]
Yes, it's a feature request.

For exaple, squeeze(), reduce_sum() an any other reduce_() function, they all have arguments where you define the axes to be operated on.

For reduce_sum(), you define which axes ought to be reduced and summed. Why not generalize the SoftMax like that too?

At the moment, the softmax API is
    tf.nn.softmax(logits, name=None)

My feature request is to to
    tf.nn.softmax(logits, softmax_dim = [1], name=None)

**Comment by Anmol6** [comment]
@RobRomijnders stupid question, but where can I find the softmax function? I wanna work on this, but don't seem to be able to find the file..

**Comment by RobRomijnders** [comment]
It's under the neural networks module, so tensorflow.nn.softmax().

Another engineer of the Google Brain team also contacted me. I think he's working on it too

**Comment by Anmol6** [comment]
@RobRomijnders Sorry, but I still can't find it... Could you please give me a link? 

Also, should I still work on it then? Perhaps, I can collaborate with the engineer? Not sure.

**Comment by vrv** [comment]
Indeed, assigned to the person who is almost done with it, although I believe what is being done here is softmax over a single but specific dimension (with the thought being that if you need to softmax over multiple contiguous dimensions, you can just reshape the adjacent dims).

**Comment by Anmol6** [comment]
@vrv ah, okay. For future reference, could you provide a link to where the softmax function is?

**Comment by vrv** [comment]
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/kernels (softmax_*) for implementations -- the python code is automatically generated from the Op specification here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/nn_ops.cc#L1261, because there's no extra logic needed in python.

**Comment by drpngx** [comment]
Superseded by #6590.

**Comment by Remi-Coulom** [comment]
I think #6590 does not address this. #210 is a duplicate, and offers that solution: https://gist.github.com/raingo/a5808fe356b8da031837

I think it would be more coherent and nicer if it could be the default behaviour of tensorflow.

**Comment by jrbtaylor** [comment]
I think this one slipped through the cracks. tf.nn.softmax still only works over 1 dimension.

**Comment by redna11** [comment]
required +1
